src/level-editor/editorState.ts

```diff
@@ -0,0 +1,113 @@
+import { createReducer, arraylike, settable } from '@dwalter/create-reducer'
+import { Resolve } from '@dwalter/spider-hook'
+import { Dispatch } from '@dwalter/spider-hook/src/types'
+
+export type Location = [number, number]
+
+export interface Tile {
+  location: Location
+  color: 'white' | 'black'
+}
+
+const [reducer, actions] = createReducer(
+  'tiles',
+  [{ location: [3, 4], color: 'white' }],
+  arraylike<Tile>(),
+)
+
+export const tilesSelector = reducer
+
+export const addTile = actions.add
+
+const [goatReducer, goatActions] = createReducer(
+  'goat',
+  [3, 4],
+  settable<Location>(),
+)
+
+export const goatLocationSelector = goatReducer
+
+export function relocateGoat(newGoat: Location) {
+  return (dispatch: Dispatch, resolve: Resolve) => {
```

Comment:
one cool thing about actions is that they can be a function. `resolve` can resolve the state inside of a selector in order to make smarter actions that read state before doing anything